spec/connection.md

```diff
@@ -0,0 +1,258 @@
+
+## Guide-level Explanation
+
+### Establishing the connection
+
+**TODO** Add WebSocket flow.
+
+#### Minimal example
+
+1. Clients initiate a network connection to the Server.
+2. Server responds with [INFO][INFO] json.
+3. Client sends [CONNECT][CONNECT] json.
+4. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+**Note** Server can send subsequent [INFO][INFO] on a ongoing connection.
+Client needs to handle them approprietly and update server lists and server info.
+
+#### Auth flow
+TODO
+
+#### TLS
+There are two flows available in the Server that enable TLS.
+
+##### Standard TLS
+
+This method is available in all NATS Server versions.
+
+1. Clients initiate a network connection to the Server.
+2. Server responds with [INFO][INFO] json.
+3. If Server [INFO][INFO] contains `tls_required` set to `true`, or the client has a tls requirement set to `true`, the client performs a TLS upgrade.
+4. Client sends [CONNECT][CONNECT] json.
+5. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+##### TLS First
+
+This method has been available since NATS Server 2.11.
+
+There are two prerequisites to use this method:
+1. Server config has enabled `handshake_first` field in the `tls` block.
+2. The client has set the `tls_first` option set to true.
+
+**handshake_first**
+has those possible values:
+- **`false`**: handshake first is disabled. Default value
+- `true`: handshake first is enabled and enforced. Clients that do not use this flow will fail to connect.
+- `duration` (i.e. 2s): a hybrid mode that will wait a given time, allowing the client to follow the `tls_first` flow. After the duration has expired, `INFO` is sent, enabling standard client TLS flow.
+- `auto`: same as above, with some default value. By default it waits 50ms for TLS upgrade before sending the [INFO][INFO].
+
+The flow itself is flipped. TLS is established before the Server sends INFO:
+
+1. Clients initiate a network connection to the Server.
+2. Client upgrades the connection to TLS.
+2. Server [INFO][INFO] json.
+4. Client sends [CONNECT][CONNECT] json.
+5. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+
+### Servers discovery
+
+When Server sends back [INFO][INFO]. It may contain additional URLs to which the client can make connection attempts.
+The client should store those URLs and use them in the Reconnection Strategy.
+
+A client should have an option to turn off using advertised URLs.
+By default, those URLs are used.
+
+**TODO**: Add more in-depth explanation how topology discovery works.
+
+### Reconnection Strategies (In progress)
+
+#### Detecting disconnection
+
+There are two methods that clients should use to detect disconnections:
+1. Missing two consecutive PONGs from the Server.
```

Comment:
The number of missed PONGs is actually configurable. It is called `MaxPingsOut` in the Go client and yes, it defaults to 2.

Reply:
Yes, it is, although I think more options we give to the users to configure reconnect logic, more footguns and potential strain to the whole system and network in can introduce. 

Understanding consequences of relation between PING interval and number of accepted PONGs missing is not trivial.
Example: Ping interval set to 1s, max pings out to 30. A lot of unnecessary chattiness with very little benefit for the responsiveness of the client.

That is why I consider if we should add that option to every new client.
One variable - Ping interval should suffice to all use cases of configuring ping behaviour, or I am missing something here?